guardian/dotcom-rendering
# Dynamic imports CMP example

## What does this change?
Implement dynamic imports using `loadable-components`  on the frontend (`react.js`) bundle.

## Why?
A proof of concept of how we can break react components out of the core `react.js` bundle.

Using `@loadable/component` to import a component will automatically split this component out of the `react.js` bundle. `loadable-components` will then fetch that component remotely if the component is trying to be rendered.

Main benefit, fetch the component if it is being used, reducing the bundle size for components that are not often used.

## Note
Q: Why are we not implementing SSR using `loadable-components`?
A: `loadable-components` SSR requires you to render the app server side and allows you to extract script tags for each component/library, giving you the opportunity to load each chunk in parallel on page load.

Our app has 2 sections, SSR-ed html and a bundled `react.js` payload. It makes no sense to do dynamic imports to our server code, as all of it gets used to generate HTML and is never used by the client. Regarding the `react.js` bundle, we would need to render it sever side and use the ChunkExtractor in order to extract script tags (refer to [step 3](https://loadable-components.com/docs/server-side-rendering/) ). However rendering the `react.js` server side would fail as we depend on the window object in several sections of the code and of course window does not exist in node.

This is something we could explore if we remove the window dependency from the bundle.

Q: why was `babel-plugin-dynamic-import-node` removed?
A: it replaced instances of `import` to `require`, thus causing issues to get dynamic imports to work

Q: why has the `CMP` component been changed into a default export?
A: Currently `loadable-components` does not support non default exports of components, there is an active PR to allow non default exports https://github.com/gregberge/loadable-components/pull/483

Q: Can this be used for other components?
A: Yes, here is a breakdown of other components that are conditionally rendered in `App.tsx`
<img width="390" alt="Screenshot 2020-04-15 at 11 04 00" src="https://user-images.githubusercontent.com/8831403/79325284-d64c1e00-7f08-11ea-9ba2-121297b81b45.png">


## How to test
You can use local dev environment to test. `CMP` is rendered conditionally using `CAPI.config.cmpUi` in `src/web/components/App.tsx`. As `CAPI.config.cmpUi` is determined on BE, you will need to replace it with a boolean `true` or `false` to see it in action.

When `CMP` is being rendered, you should be able to see an import in the network tab. When it is not, you should not see any `CMP` call in the network tab.

## Link to supporting Trello card
https://trello.com/c/0qCpwbbO/1285-dynamic-imports
